test/cpp/api/inference_mode.cpp

```diff
@@ -0,0 +1,276 @@
+  ASSERT_EQ(c10::impl::is_all_dispatch_keyset_excluded(c10::autograd_dispatch_keyset), inference_mode);
+  ASSERT_EQ(c10::impl::tls_is_dispatch_key_included(c10::DispatchKey::InplaceOrView), !inference_mode);
+}
+
+TEST(InferenceModeTest, TestTLSState) {
+  assert_TLS_states(false);
+  {
+    InferenceMode guard;
+    assert_TLS_states(true);
+    {
+      InferenceMode guard(false);
+      assert_TLS_states(false);
+    }
+    assert_TLS_states(true);
+  }
+  assert_TLS_states(false);
+}
+
+TEST(InferenceModeTest, TestInferenceTensorCreation) {
+  {
+    InferenceMode guard;
+    // New tensor created in InferenceMode (inference tensor) doesn't have Autograd & InplaceOrView keys.
+    // Even if you ask it to requires_grad (which doesn't make sense in InferenceMode),
+    // it'll be silently ignored. This is not an error to make adding InferenceMode
+    // guard to existing code a smooth user experience.
+    torch::Tensor c = torch::ones({1, 2, 3});
+    assert_tensor_dispatch_keys(c, /*is_inference_tensor=*/true);
+    ASSERT_FALSE(c.requires_grad());
+
+    torch::Tensor tmp = torch::ones({1, 2, 3}).set_requires_grad(true);
+    ASSERT_TRUE(tmp.requires_grad()); // requires_grad is silently ignored when it's an inference tensor.
+    assert_tensor_dispatch_keys(tmp, /*is_inference_tensor=*/true);
+
+    tmp = torch::ones({1, 2, 3}).set_requires_grad(false);
+    ASSERT_FALSE(tmp.requires_grad());
+    assert_tensor_dispatch_keys(tmp, /*is_inference_tensor=*/true);
+  }
+}
+
+TEST(InferenceModeTest, TestExistingAutogradSession) {
+  torch::Tensor s = torch::ones({1, 2, 3}).set_requires_grad(true);
+  torch::Tensor a = s + 2;
+
+  // Save `a` in an existing autograd session
+  torch::Tensor out = a * a;
+  {
+    InferenceMode guard;
+    inplace_op(a);
+  }
+  // perform backward on `a` should trigger error since `a`'s version has been bumped.
+  ASSERT_THROWS_WITH(out.backward(torch::ones_like(out)),
+    "one of the variables needed for gradient computation has been modified by an inplace operation")
+}
+
+TEST(InferenceModeTest, TestInferenceTensorInInferenceMode) {
+  c10::InferenceMode guard;
+  torch::Tensor c = torch::ones({1, 2, 3});
+
+  torch::Tensor func_out = functional_op(c);  // go through kernels: CPU
+  assert_tensor_dispatch_keys(func_out, /*is_inference_tensor=*/true);
+
+  inplace_op(c);  // go through kernels: CPU
+  assert_tensor_dispatch_keys(c, /*is_inference_tensor=*/true);
+
+  torch::Tensor view_out = view_op(c);  // go through kernels: CPU
+  assert_tensor_dispatch_keys(view_out, /*is_inference_tensor=*/true);
+}
+
+TEST(InferenceModeTest, TestInferenceTensorInNormalMode) {
+  torch::Tensor inference_tensor;
+  {
+    InferenceMode guard;
+    inference_tensor = torch::ones({1, 2, 3});
+  }
+  // We cannot add c10::autograd_dispatch_keyset to gloablly enabled, because that will
+  // accidentally call autograd kernel from a backend that might not match tensor input.
```

Comment:
I don't really understand this sentence. Is it setting up some sort of counterfactual? *Suppose* we added autograd dispatch keyset to globally enabled (word missing here), then... this test would fail? The sentence seems like it is making a statement about how AutogradXLA/AutogradCPU key design means you can't put them in the global set, but that doesn't seem relevant to the test here?